### PR TITLE
fix node discover test case issue: RackHD/RackHD issue #270

### PR DESCRIPTION
**Pull request by changev**
enhance https://github.com/RackHD/RackHD/issues/270

Import a external value **NODES_NUMBER** ,which means how many nodes are used for test.

Enhanced node discovery test case skip condition, only if all nodes are in mongodb this case skip.

@iceiilin @panpan0000 @RackHD/corecommitters    

Jenkins: depends on https://github.com/RackHD/on-build-config/pull/23

### CODE REVIEW

**Comment by changev** [comment]
test this please

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by changev** [comment]
test this please

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by changev** [comment]
test this please

**Comment by iceiilin** [comment]
It is robust to cover the corner case when nodes are being discovered and not ready for further catalog test. :+1: 
@jlongever

**Comment by changev** [comment]
test this please

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by jlongever** [comment]
It seems like a lot of changes to add a fixed node count option.  Could you do something similar to this: https://github.com/jlongever/RackHD/commit/acda755022920ba4ed90313a3a5cd176b6e70357 ?

Also keep in mind that with a fixed node count all of the nodes will have to be deleted once you get to the next 2.0 test group, otherwise you'll always timeout.  Because this change creates a hard expectation to wait for `n` number of discovery graphs to complete for each API node discovery test.  Maybe add a dependent test to clear out all node documents prior to running the `node.discovery.test` if `RACKHD_NODES_NUMBER` is defined?

**Comment by rackhdmirror** [comment]
Build finished.

File: test/tests/api/v1_1/nodes_tests.py [hunk 1]
```diff
@@ -92,21 +94,28 @@ def __post_workflow(self, id, graph_name, data):
...
-        count = 0
+        discovered_count = 0
         for n in nodes:
             type = n.get('type')
             if type == 'compute':
-                count += 1
-        return count
+                id = n.get('id')
+                status = self.__get_workflow_status(id)
+                if status != 'pending' and status != 'running':
```
Comment (jlongever) [inline test/tests/api/v1_1/nodes_tests.py]: This is probably not safe.   You're assuming the status will be correct which we have seen in the past its not always the case.  The status in AMQP is probably more reliable.  Secondly,  it assumes the status is for discovery graph,  other tests run workflows against nodes.  so by the time you get to 2.0 test groups,  you could be looking at workflow status for other workflows executed in prior tests.   It would probably be safer to just go through each node and check if the catalog is fulfilled.

Comment (changev) [inline test/tests/api/v1_1/nodes_tests.py]: Thanks, I got it. 

So in V1_1 and V2_0 tests, instead of "find in db and no related active workflow", we should check catalogs and make sure a node is actually discovered.

"to_be_discovered" is still useful, right?

Comment (jlongever) [inline test/tests/api/v1_1/nodes_tests.py]: IMHO, I think the logic is a bit unnecessary.  Something like https://github.com/jlongever/RackHD/commit/acda755022920ba4ed90313a3a5cd176b6e70357 but include node catalog checks if the test is using a fixed count?

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by changev** [comment]
test this please

**Comment by rackhdmirror** [comment]
Build finished.

File: test/tests/api/v1_1/nodes_tests.py [hunk 2]
```diff
@@ -106,9 +106,11 @@ def get_undiscovered_compute_count(self):
                 workflows = self.__get_data()
                 LOG.info(workflows)
                 for workflow in workflows:
-                    if workflow.get['injectableName'] == 'Graph.SKU.Discovery' \
-                     and workflow.get['status'] == 'succeeded':
-                     discovered_count += 1
+                    definition = workflow.get('definition', {})
+                    injectableName = definition.get('injectableName')
+                    if injectableName == 'Graph.SKU.Discovery':
+                        workflow.get['_status'] == 'succeeded':
```
Comment (jlongever) [inline test/tests/api/v1_1/nodes_tests.py]: As I mentioned before,  this isn't going to be reliable.

Comment (changev) [inline test/tests/api/v1_1/nodes_tests.py]: a succeeded Graph.SKU.Discovery workflow doesn't means the node has been discovered successfully?

Comment (changev) [inline test/tests/api/v1_1/nodes_tests.py]: yeah i see "You're assuming the status will be correct which we have seen in the past its not always the case, The status in AMQP is probably more reliable" But why it's not reliable

**Comment by jlongever** [comment]
@changev Sorry,  I think this PR is a -1 for me.  It seems reasonable to add an optional fixed node count as an enhancement,  but I'm not quite sure what "bug" these changes are fixing.  The random node discovery failures are more likely environment based (e.g. nodes failing to PXE boot,  or missed `graph.finished` AMQP events, etc) than they are test logic based.

Fwiw,  the node discovery tests haven't changed much,  and for the most part when we see random discovery failures,  its due to a real regression or test env change.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by changev** [comment]
Ok ,thanks joel.  I realize that you haven't known the bug clearly yet.
The "bug" has shown clearly in related issue. When add more vnodes for testing, especially add physic nodes(spend more time for get ip and discover), this issue **is almost 100% occurs.** 
**Before all nodes up, some nodes has been write to mongodb and all nodes are seen as discovered ,the discovery test skip , then tests failed**. This is indeed a wrong logic.

1 I think the key now is why "a succeeded Graph.SKU.Discovery workflow" not reliable.
2 If it's true not reliable, is there any other solution you suggested can make sure a node has been discovered? Catalogs doesn't has a "discovered field" and different vnodes has different kinds of catalogs so we can't count them.

**Comment by jlongever** [comment]
@changev As I mentioned,  adding an optional fixed node count helps with boot time variances when using different systems (virtual & physical).  But that is only an enhancement since the current env doesn't mix systems. 

1) interrogate the rackhd logs, and the environment when we see a random node failure to root cause (See my comment in https://github.com/RackHD/RackHD/issues/270)
2) the proper way to determine if the discovery graph completed,  is to listen for `graph.finished` events on AMQP.  The test already does that.  We have seen that polling the /workflows is less reliable because the `status` is not always updated in mongo when it is checked.

**Comment by zyoung51** [comment]
@changev Based on your comment above this is a -1 for me too.  The solution to the test problem should not be to ignore "unreliable" things in the code under test and the root cause should be understood before the fix is submitted.  What root caused issue are you trying to fix by using a hard coded node count?

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by changev** [comment]
@jlongever 
```
count = defaults.get('RACKHD_NODE_COUNT', '')
if count.isdigit() and self.check_compute_count() == int(count):`
```

self.check_compute_count() == int(count) is enough to solve the time issue.

**Comment by jlongever** [comment]
@changev it would be safer to check `isdigit()` as well.

**Comment by changev** [comment]
@jlongever 
already checked, thanks

**Comment by jlongever** [comment]
@changev thanks,  missed the commit update.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by rackhdmirror** [comment]
Build finished.

**Comment by changev** [comment]
test this please

**Comment by cgx027** [comment]
looks good to me. :+1:

**Comment by yyscamper** [comment]
@jlongever @zyoung51 : bases on the latest change, do you still have any concern for this PR?
